File: AocCsharp/Day12/Day12.cs
using AocCsharp.Utils;

namespace AocCsharp.Day12;

public static class Day12
{
    public static void Run()
    {
        var testInputPath = Path.Combine("..", "inputs", "day12_test.txt");
        var realInputPath = Path.Combine("..", "inputs", "day12.txt");

        TestRunner.RunSolution("Part 1", Part1, testInputPath, realInputPath, expectedTestResult: 2);
        Console.WriteLine("\n🎄 Part 2 automatically completed! Both stars earned! 🎄\n");
    }

    public static int Part1(string input)
    {
        var (shapes, regions) = ParseInput(input);

        // Precompute all shape orientations once
        var allOrientations = new Shape[shapes.Count][];
        for (int i = 0; i < shapes.Count; i++)
        {
            allOrientations[i] = GetAllOrientations(shapes[i]);
        }

        int count = 0;
        foreach (var region in regions)
        {
            if (CanFitAllPresents(region, allOrientations))
            {
                count++;
            }
        }

        return count;
    }

    private static (List<List<string>> shapes, List<Region> regions) ParseInput(string input)
    {
        var lines = input.Replace("\r", "").Split('\n');
        var shapes = new List<List<string>>();
        var regions = new List<Region>();

        int i = 0;
        while (i < lines.Length)
        {
            var line = lines[i];

            // Parse region (check this first since it also contains ':')
            if (line.Contains('x') && line.Contains(':'))
            {
                var parts = line.Split(':');
                var dimensions = parts[0].Trim().Split('x');
                int width = int.Parse(dimensions[0]);
                int height = int.Parse(dimensions[1]);

                var countParts = parts[1].Trim().Split(' ', StringSplitOptions.RemoveEmptyEntries);
                var counts = new int[countParts.Length];
                for (int j = 0; j < countParts.Length; j++)
                {
                    counts[j] = int.Parse(countParts[j]);
                }

                regions.Add(new Region(width, height, counts));
                i++;
            }
            // Parse shape
            else if (line.Contains(':'))
            {
                var shapeLines = new List<string>();
                i++; // Skip the label line

                while (i < lines.Length && !string.IsNullOrWhiteSpace(lines[i]) && !lines[i].Contains(':'))
                {
                    shapeLines.Add(lines[i]);
                    i++;
                }

                if (shapeLines.Count > 0)
                {
                    shapes.Add(shapeLines);
                }
            }
            else
            {
                i++;
            }
        }

        return (shapes, regions);
    }

    private static bool CanFitAllPresents(Region region, Shape[][] allOrientations)
    {
        // Quick area check
        int totalArea = region.Width * region.Height;
        int requiredArea = 0;

        for (int i = 0; i < region.Counts.Length; i++)
        {
            if (region.Counts[i] > 0)
            {
                int shapeArea = allOrientations[i][0].Area;
                requiredArea += shapeArea * region.Counts[i];
            }
        }

        if (requiredArea > totalArea)
        {
            return false;
        }

        var grid = new bool[region.Height * region.Width];
        var counts = new int[region.Counts.Length];
        Array.Copy(region.Counts, counts, region.Counts.Length);

        return TryPlacePresents(grid, region.Width, region.Height, counts, allOrientations, 0);
    }

    private static bool TryPlacePresents(bool[] grid, int width, int height, int[] counts,
        Shape[][] allOrientations, int callCount)
    {
        if (callCount > 2000000) return false;

        // Check if all presents are placed
        bool allPlaced = true;
        for (int i = 0; i < counts.Length; i++)
        {
            if (counts[i] > 0)
            {
                allPlaced = false;
                break;
            }
        }
        if (allPlaced) return true;

        // Try placing first available present type
        for (int shapeIdx = 0; shapeIdx < counts.Length; shapeIdx++)
        {
            if (counts[shapeIdx] == 0) continue;

            var orientations = allOrientations[shapeIdx];

            // Try each orientation at each position
            foreach (var shape in orientations)
            {
                for (int row = 0; row <= height - shape.Height; row++)
                {
                    for (int col = 0; col <= width - shape.Width; col++)
                    {
                        if (CanPlaceShape(grid, width, height, shape, row, col))
                        {
                            PlaceShape(grid, width, shape, row, col);
                            counts[shapeIdx]--;

                            if (TryPlacePresents(grid, width, height, counts, allOrientations, callCount + 1))
                            {
                                return true;
                            }

                            RemoveShape(grid, width, shape, row, col);
                            counts[shapeIdx]++;
                        }
                    }
                }
            }

            // If we couldn't place this present type anywhere, fail
            return false;
        }

        return true;
    }

    private static bool CanPlaceShape(bool[] grid, int width, int height, Shape shape, int startRow, int startCol)
    {
        if (startRow + shape.Height > height) return false;
        if (startCol + shape.Width > width) return false;

        for (int i = 0; i < shape.CellCount; i++)
        {
            int r = shape.Rows[i];
            int c = shape.Cols[i];
            int gridIdx = (startRow + r) * width + (startCol + c);

            if (grid[gridIdx])
            {
                return false;
            }
        }

        return true;
    }

    private static void PlaceShape(bool[] grid, int width, Shape shape, int startRow, int startCol)
    {
        for (int i = 0; i < shape.CellCount; i++)
        {
            int r = shape.Rows[i];
            int c = shape.Cols[i];
            int gridIdx = (startRow + r) * width + (startCol + c);
            grid[gridIdx] = true;
        }
    }

    private static void RemoveShape(bool[] grid, int width, Shape shape, int startRow, int startCol)
    {
        for (int i = 0; i < shape.CellCount; i++)
        {
            int r = shape.Rows[i];
            int c = shape.Cols[i];
            int gridIdx = (startRow + r) * width + (startCol + c);
            grid[gridIdx] = false;
        }
    }

    private static Shape[] GetAllOrientations(List<string> shapeTemplate)
    {
        var orientations = new HashSet<string>();
        var current = shapeTemplate;

        for (int rotation = 0; rotation < 4; rotation++)
        {
            var key = string.Join("|", current);
            orientations.Add(key);

            var flipped = Flip(current);
            var flippedKey = string.Join("|", flipped);
            orientations.Add(flippedKey);

            current = Rotate(current);
        }

        var result = new Shape[orientations.Count];
        int idx = 0;
        foreach (var s in orientations)
        {
            result[idx++] = new Shape(s.Split('|'));
        }
        return result;
    }

    private static List<string> Rotate(List<string> shape)
    {
        int rows = shape.Count;
        int cols = shape[0].Length;
        var rotated = new List<string>();

        for (int c = 0; c < cols; c++)
        {
            var newRow = new char[rows];
            for (int r = 0; r < rows; r++)
            {
                newRow[r] = shape[rows - 1 - r][c];
            }
            rotated.Add(new string(newRow));
        }

        return rotated;
    }

    private static List<string> Flip(List<string> shape)
    {
        var flipped = new List<string>(shape.Count);
        for (int i = 0; i < shape.Count; i++)
        {
            var row = shape[i];
            var chars = new char[row.Length];
            for (int j = 0; j < row.Length; j++)
            {
                chars[j] = row[row.Length - 1 - j];
            }
            flipped.Add(new string(chars));
        }
        return flipped;
    }

    private record Region(int Width, int Height, int[] Counts);

    private class Shape
    {
        public int[] Rows;
        public int[] Cols;
        public int CellCount;
        public int Width;
        public int Height;
        public int Area;

        public Shape(string[] lines)
        {
            var cells = new List<(int row, int col)>();

            for (int r = 0; r < lines.Length; r++)
            {
                for (int c = 0; c < lines[r].Length; c++)
                {
                    if (lines[r][c] == '#')
                    {
                        cells.Add((r, c));
                    }
                }
            }

            CellCount = cells.Count;
            Rows = new int[CellCount];
            Cols = new int[CellCount];

            for (int i = 0; i < CellCount; i++)
            {
                Rows[i] = cells[i].row;
                Cols[i] = cells[i].col;
            }

            Height = lines.Length;
            Width = 0;
            for (int i = 0; i < lines.Length; i++)
            {
                if (lines[i].Length > Width) Width = lines[i].Length;
            }
            Area = CellCount;
        }
    }
}

File: AocC/day12.c
#include "day12.h"
#include "runner.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

// ANSI color codes
#define COLOR_RESET   "\033[0m"
#define COLOR_GREEN   "\033[32m"

void day12_run(void) {
    const char* test_path = "../inputs/day12_test.txt";
    const char* real_path = "../inputs/day12.txt";

    run_solution("Part 1", day12_part1, test_path, real_path, 2);
    printf("\n" COLOR_GREEN "🎄 Part 2 automatically completed! Both stars earned! 🎄" COLOR_RESET "\n");
}

#define MAX_SHAPES 10
#define MAX_CELLS 20
#define MAX_ORIENTATIONS 8
#define MAX_REGIONS 100

typedef struct {
    int rows[MAX_CELLS];
    int cols[MAX_CELLS];
    int cell_count;
    int width;
    int height;
} Shape;

typedef struct {
    int width;
    int height;
    int counts[MAX_SHAPES];
    int num_counts;
} Region;

static void normalize_shape(int* rows, int* cols, int n, Shape* out) {
    if (n == 0) {
        out->cell_count = 0;
        out->width = 0;
        out->height = 0;
        return;
    }

    int min_r = rows[0], min_c = cols[0];
    for (int i = 1; i < n; i++) {
        if (rows[i] < min_r) min_r = rows[i];
        if (cols[i] < min_c) min_c = cols[i];
    }

    int max_r = 0, max_c = 0;
    for (int i = 0; i < n; i++) {
        out->rows[i] = rows[i] - min_r;
        out->cols[i] = cols[i] - min_c;
        if (out->rows[i] > max_r) max_r = out->rows[i];
        if (out->cols[i] > max_c) max_c = out->cols[i];
    }

    out->cell_count = n;
    out->height = max_r + 1;
    out->width = max_c + 1;
}

static void rotate_points(int* rows, int* cols, int n, int* out_rows, int* out_cols) {
    for (int i = 0; i < n; i++) {
        out_rows[i] = cols[i];
        out_cols[i] = -rows[i];
    }
}

static void flip_points(int* rows, int* cols, int n, int* out_rows, int* out_cols) {
    for (int i = 0; i < n; i++) {
        out_rows[i] = rows[i];
        out_cols[i] = -cols[i];
    }
}

static void shape_to_key(Shape* s, char* key) {
    // Create a grid representation as a string key
    char grid[20][20];
    memset(grid, '.', sizeof(grid));

    for (int i = 0; i < s->cell_count; i++) {
        grid[s->rows[i]][s->cols[i]] = '#';
    }

    int pos = 0;
    for (int r = 0; r < s->height; r++) {
        for (int c = 0; c < s->width; c++) {
            key[pos++] = grid[r][c];
        }
        key[pos++] = '|';
    }
    key[pos] = '\0';
}

static int get_orientations(int* orig_rows, int* orig_cols, int n, Shape orientations[MAX_ORIENTATIONS]) {
    int count = 0;
    char seen_keys[MAX_ORIENTATIONS][512];

    int curr_rows[MAX_CELLS], curr_cols[MAX_CELLS];
    memcpy(curr_rows, orig_rows, n * sizeof(int));
    memcpy(curr_cols, orig_cols, n * sizeof(int));

    for (int rot = 0; rot < 4; rot++) {
        // Normal
        Shape s;
        normalize_shape(curr_rows, curr_cols, n, &s);
        char key[512];
        shape_to_key(&s, key);

        bool found = false;
        for (int i = 0; i < count; i++) {
            if (strcmp(seen_keys[i], key) == 0) {
                found = true;
                break;
            }
        }
        if (!found && count < MAX_ORIENTATIONS) {
            strcpy(seen_keys[count], key);
            orientations[count++] = s;
        }

        // Flipped
        int flip_rows[MAX_CELLS], flip_cols[MAX_CELLS];
        flip_points(curr_rows, curr_cols, n, flip_rows, flip_cols);
        normalize_shape(flip_rows, flip_cols, n, &s);
        shape_to_key(&s, key);

        found = false;
        for (int i = 0; i < count; i++) {
            if (strcmp(seen_keys[i], key) == 0) {
                found = true;
                break;
            }
        }
        if (!found && count < MAX_ORIENTATIONS) {
            strcpy(seen_keys[count], key);
            orientations[count++] = s;
        }

        // Rotate for next iteration
        int rot_rows[MAX_CELLS], rot_cols[MAX_CELLS];
        rotate_points(curr_rows, curr_cols, n, rot_rows, rot_cols);
        memcpy(curr_rows, rot_rows, n * sizeof(int));
        memcpy(curr_cols, rot_cols, n * sizeof(int));
    }

    return count;
}

static bool can_place(bool* grid, int grid_w, int grid_h, Shape* s, int start_r, int start_c) {
    if (start_r + s->height > grid_h) return false;
    if (start_c + s->width > grid_w) return false;

    for (int i = 0; i < s->cell_count; i++) {
        int r = start_r + s->rows[i];
        int c = start_c + s->cols[i];
        if (grid[r * grid_w + c]) return false;
    }
    return true;
}

static void place_shape(bool* grid, int grid_w, Shape* s, int start_r, int start_c, bool val) {
    for (int i = 0; i < s->cell_count; i++) {
        int r = start_r + s->rows[i];
        int c = start_c + s->cols[i];
        grid[r * grid_w + c] = val;
    }
}

static bool solve(bool* grid, int w, int h, int* counts, int num_shapes,
                  Shape orientations[MAX_SHAPES][MAX_ORIENTATIONS],
                  int orient_counts[MAX_SHAPES]) {
    // Find first shape with count > 0
    int shape_idx = -1;
    for (int i = 0; i < num_shapes; i++) {
        if (counts[i] > 0) {
            shape_idx = i;
            break;
        }
    }

    if (shape_idx == -1) return true;  // All placed

    for (int o = 0; o < orient_counts[shape_idx]; o++) {
        Shape* s = &orientations[shape_idx][o];

        for (int r = 0; r <= h - s->height; r++) {
            for (int c = 0; c <= w - s->width; c++) {
                if (can_place(grid, w, h, s, r, c)) {
                    place_shape(grid, w, s, r, c, true);
                    counts[shape_idx]--;

                    if (solve(grid, w, h, counts, num_shapes, orientations, orient_counts)) {
                        return true;
                    }

                    counts[shape_idx]++;
                    place_shape(grid, w, s, r, c, false);
                }
            }
        }
    }

    return false;
}

int64_t day12_part1(const char* input) {
    char* data = strdup(input);
    if (!data) return 0;

    // Parse shapes and regions
    int shape_rows[MAX_SHAPES][MAX_CELLS];
    int shape_cols[MAX_SHAPES][MAX_CELLS];
    int shape_sizes[MAX_SHAPES];
    int num_shapes = 0;

    Region regions[MAX_REGIONS];
    int num_regions = 0;

    char* lines[500];
    int line_count = 0;

    char* line = strtok(data, "\r\n");
    while (line && line_count < 500) {
        lines[line_count++] = line;
        line = strtok(NULL, "\r\n");
    }

    int i = 0;
    while (i < line_count) {
        char* l = lines[i];

        // Check if it's a region definition (contains 'x' and ':')
        if (strchr(l, 'x') && strchr(l, ':')) {
            Region r = {0};
            char* colon = strchr(l, ':');
            *colon = '\0';

            sscanf(l, "%dx%d", &r.width, &r.height);

            char* counts_str = colon + 1;
            char* tok = strtok(counts_str, " \t");
            while (tok && r.num_counts < MAX_SHAPES) {
                r.counts[r.num_counts++] = atoi(tok);
                tok = strtok(NULL, " \t");
            }

            if (num_regions < MAX_REGIONS) {
                regions[num_regions++] = r;
            }
            i++;
        }
        // Check if it's a shape definition (contains ':' but not 'x')
        else if (strchr(l, ':')) {
            i++;  // Skip label

            // Read shape lines
            int rows[MAX_CELLS], cols[MAX_CELLS];
            int n = 0;
            int row = 0;

            while (i < line_count && strlen(lines[i]) > 0 && !strchr(lines[i], ':')) {
                for (int c = 0; c < (int)strlen(lines[i]) && n < MAX_CELLS; c++) {
                    if (lines[i][c] == '#') {
                        rows[n] = row;
                        cols[n] = c;
                        n++;
                    }
                }
                row++;
                i++;
            }

            if (n > 0 && num_shapes < MAX_SHAPES) {
                memcpy(shape_rows[num_shapes], rows, n * sizeof(int));
                memcpy(shape_cols[num_shapes], cols, n * sizeof(int));
                shape_sizes[num_shapes] = n;
                num_shapes++;
            }
        } else {
            i++;
        }
    }

    // Precompute orientations
    Shape orientations[MAX_SHAPES][MAX_ORIENTATIONS];
    int orient_counts[MAX_SHAPES];

    for (int s = 0; s < num_shapes; s++) {
        orient_counts[s] = get_orientations(shape_rows[s], shape_cols[s],
                                             shape_sizes[s], orientations[s]);
    }

    // Try each region
    int64_t valid_count = 0;

    for (int r = 0; r < num_regions; r++) {
        Region* reg = &regions[r];

        // Quick area check
        int required_area = 0;
        for (int s = 0; s < reg->num_counts && s < num_shapes; s++) {
            required_area += reg->counts[s] * shape_sizes[s];
        }
        if (required_area > reg->width * reg->height) continue;

        // Create grid
        bool* grid = (bool*)calloc(reg->width * reg->height, sizeof(bool));
        int counts[MAX_SHAPES];
        memcpy(counts, reg->counts, reg->num_counts * sizeof(int));

        if (solve(grid, reg->width, reg->height, counts, reg->num_counts,
                  orientations, orient_counts)) {
            valid_count++;
        }

        free(grid);
    }

    free(data);
    return valid_count;
}
